### PR TITLE
[DeviceASAN] Use device usm to sync asan runtime data instead of shared usm

### PR DESCRIPTION
**Pull request by zhaomaosu**
Shared USM has poor performance, change it to device USM will benefit several benchmarks.

intel/llvm part: https://github.com/intel/llvm/pull/15911

### CODE REVIEW

**Review by yingcong-wu — APPROVED** [review]

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 1]
```diff
@@ -682,6 +663,34 @@ ur_result_t SanitizerInterceptor::prepareLaunch(
...
+        LaunchInfo.Data.Host.Debug = getOptions().Debug ? 1 : 0;
+
+        UR_CALL(getContext()->urDdiTable.USM.pfnDeviceAlloc(
+            ContextInfo->Handle, DeviceInfo->Handle, nullptr, nullptr,
+            sizeof(LaunchInfo), (void **)&LaunchInfo.Data.DevicePtr));
+        getContext()->logger.debug(
+            "launch_info {} (numLocalArgs={}, localArgs={})",
+            (void *)LaunchInfo.Data.DevicePtr,
+            LaunchInfo.Data.Host.NumLocalArgs,
+            (void *)LaunchInfo.Data.Host.LocalArgs);
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:686]: no host-device sync?

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 2]
```diff
@@ -154,8 +154,50 @@ struct ContextInfo {
     }
 };
 
-struct USMLaunchInfo {
-    LaunchInfo *Data = nullptr;
+struct AsanRuntimeDataWrapper {
+    AsanRuntimeData Host{};
+
+    AsanRuntimeData *DevicePtr = nullptr;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.hpp:160]: please manage this pointer in AsanRuntimeDataWrapper

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp:160]: Done

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 1 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Refined this part.

**Comment by pbalcer** [comment]
this is a non-trivial change, please create an intel/llvm PR.

**Comment by zhaomaosu** [comment]
> this is a non-trivial change, please create an intel/llvm PR.

Done

**Review by pbalcer — APPROVED** [review]

**Review by AllanZyne — APPROVED** [review]

**Comment by AllanZyne** [comment]
please rebase.

**Comment by zhaomaosu** [comment]
> This needs a rebase again.

Done

**Comment by zhaomaosu** [comment]
The build failure is fixed in intel/llvm part PR. And the pre ci of intel/llvm PR is passed.

**Pull request merged**